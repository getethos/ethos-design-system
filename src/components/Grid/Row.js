import PropTypes from 'prop-types'
import React from 'react'
import styles from './Row.module.scss'

export const Row = React.memo((props) => {
  let css
  if (props.size !== Row.SIZES.DEFAULT) {
    css = styles[`row${props.size}`]
  } else {
    css = props.className
  }

  return (
    <div className={css} role="row">
      {React.Children.map(props.children, (column, index) =>
        // From https://bit.ly/2tq3FiS
        // ...`props.children` isn't the actual children; It is the descriptor
        // of the children. So you don't have actually anything to change; you
        // can't change any props, or edit any functionality; you can only read
        // from it. If you need to make any modifications you have to create new
        // elements using React.CloneElement.
        React.cloneElement(column, {
          active: props.active && index === props.columnIndex,
        })
      )}
    </div>
  )
})

Row.displayName = 'Row'

Row.SIZES = {
  SMALL: 'Small',
  DEFAULT: 'Default',
  LARGE: 'Large',
}

Row.propTypes = {
  size: PropTypes.oneOf(Object.values(Row.SIZES)),
  active: PropTypes.bool,
  columnIndex: PropTypes.number,
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
}

Row.defaultProps = {
  size: Row.SIZES.DEFAULT,
  active: false,
  columnIndex: -1,
  className: styles.row,
}
