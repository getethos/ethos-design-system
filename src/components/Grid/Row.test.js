import React from 'react'
import TestRenderer from 'react-test-renderer'

import { Row } from './Row.js'

function render(props) {
  return TestRenderer.create(
    <Row {...props}>
      <div>I am the first column</div>
      <div>I am the second column</div>
    </Row>
  )
}

test('renders the default state', () => {
  const row = TestRenderer.create(
    <Row>
      <div>Foo</div>
      <div>Bar</div>
      <div>Baz</div>
    </Row>
  ).toJSON()
  expect(row).toMatchSnapshot()
})

test('className', () => {
  const row = TestRenderer.create(
    <Row className="fooby">
      <div>Foo</div>
      <div>Bar</div>
      <div>Baz</div>
    </Row>
  ).toJSON()
  expect(row).toMatchSnapshot()
  expect(row.props.className).toStrictEqual('fooby')
})

test('columns are inactive be default', () => {
  const row = TestRenderer.create(
    <Row active={false} columnIndex={-1}>
      <div>Foo</div>
      <div>Bar</div>
    </Row>
  ).toJSON()
  expect(row.children[0].props.active).toStrictEqual(false)
  expect(row.children[1].props.active).toStrictEqual(false)
})

test('adds active attribute to column index specified', () => {
  for (let i = 0; i < 2; i++) {
    const renderer = render({ active: true, columnIndex: i })
    const root = renderer.toJSON()
    expect(root.children[i].props.active).toStrictEqual(true)
  }
})
