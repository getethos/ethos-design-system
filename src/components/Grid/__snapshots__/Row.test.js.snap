// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`className 1`] = `
<div
  className="fooby"
  role="row"
>
  <div
    active={false}
  >
    Foo
  </div>
  <div
    active={false}
  >
    Bar
  </div>
  <div
    active={false}
  >
    Baz
  </div>
</div>
`;

exports[`renders the default state 1`] = `
<div
  className="row"
  role="row"
>
  <div
    active={false}
  >
    Foo
  </div>
  <div
    active={false}
  >
    Bar
  </div>
  <div
    active={false}
  >
    Baz
  </div>
</div>
`;
